io.c

```diff
@@ -784,6 +784,87 @@ static int io_write(struct io_kiocb *req, const struct sqe_submit *s,
 	return ret;
 }
 
+static int io_discard(struct io_kiocb *req, const struct sqe_submit *s,
+	bool force_nonblock)
+{
+	const int mode = FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE;
+	struct file *file = req->file;
+	struct inode *inode = file->f_mapping->host;
+	int ret = -EINVAL;
+	loff_t pos = s->pos;
+	size_t bytes = s->len;
+	
+	/* discard always requires a blocking context */
+	if (force_nonblock)
+		return -EAGAIN;
+
+	if (unlikely(!(req->file->f_mode & FMODE_WRITE)))
+		return -EBADF;
+
+	if (S_ISREG(inode->i_mode)) {
+		if (unlikely(!req->file->f_op->fallocate)) {
+			pr_info("%s: fallocate is NULL", __func__);
+			return -EOPNOTSUPP;
+		}
+
+		ret = req->file->f_op->fallocate(req->file, mode, pos, bytes);
+		if (unlikely(ret && ret != -EINVAL && ret != -EOPNOTSUPP))
+			ret = -EIO;
+	} else  if (S_ISBLK(inode->i_mode)) {
+		struct block_device *bdev = I_BDEV(inode);
+		sector_t start_sector = (pos >> SECTOR_SHIFT);
+		sector_t nsectors = sector_div(bytes, SECTOR_SIZE);
+
+		// santize byte range for sector granularity
+		if ((pos & (SECTOR_SIZE-1)) || (nsectors & (SECTOR_SIZE-1))) {
+			return -EINVAL;
+		}
+		ret = blkdev_issue_discard(bdev, start_sector, nsectors, GFP_KERNEL, 0);
```

Comment:
I think in 4.x+ kernels the block devices support fallocate. uring is supported from 4.15, so you might be able to simply call fallocate all the time.